Improve the content for `read_lines` (#1679)
rust-lang/rust-by-example: src/std_misc/file/read_lines.md

# `read_lines`

## A naive approach

This might be a reasonable first attempt for a beginner's first
implementation for reading lines from a file.

```rust,norun
use std::fs::read_to_string;

fn read_lines(filename: &str) -> Vec<String> {
    let mut result = Vec::new();

    for line in read_to_string(filename).unwrap().lines() {
        result.push(line.to_string())
    }

    result
}
```

Since the method `lines()` returns an iterator over the lines in the file,
we can also perform a map inline and collect the results, yielding a more
concise and fluent expression.

```rust,norun
use std::fs::read_to_string;

fn read_lines(filename: &str) -> Vec<String> {
    read_to_string(filename) 
        .unwrap()  // panic on possible file-reading errors
        .lines()  // split the string into an iterator of string slices
        .map(String::from)  // make each slice into a string
        .collect()  // gather them together into a vector
}
```

Note that in both examples above, we must convert the `&str` reference
returned from `lines()` to the owned type `String`, using `.to_string()`
and `String::from` respectively.

## A more efficient approach

> Here we use the `BufRead` class to read the file. `BufRead` uses an internal

Reviewer: ```suggestion
Here we pass ownership of the open `File` to a `BufReader` struct. `BufReader` uses an internal
```

> readahead buffer to performantly reduce the number of times the underlying
> file storage layer must be queried per line read. The intermediary buffer
> is much faster to access.

Reviewer: ```suggestion
buffer to reduce intermediate allocations.
```
